### PR TITLE
Release new version

### PR DESCRIPTION
**Pull request by ailox**
A lot of useful changes have been implemented, it's time to make them available!

### CODE REVIEW

File: ubirch/keystore.go [hunk 1]
```diff
@@ -61,6 +68,24 @@ func (enc *EncryptedKeystore) SetKey(keyname string, keyvalue []byte) error {
...
+		return nil, err
+	}
+	kek, err := u.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	return enc.Keystore.Get(keyname2, kek)
+}
+
```
Comment (ailox) [inline ubirch/keystore.go:88]: This shouldn't be in here

Comment (lohrke) [reply ubirch/keystore.go]: Fixed by commit https://github.com/ubirch/ubirch-protocol-go/pull/18/commits/53f145718cbf406edb6f1f7d73dac86eaacb731b

File: ubirch/test_benchmark_common_test.go [hunk 2]
```diff
@@ -244,3 +308,142 @@ func verifyUPPChain(t *testing.T, uppsArray [][]byte, startSignature []byte) err
...
+	//... check chain iself
+	err = verifyUPPChain(t, uppsArray, lastSigBytes)
+	if err != nil {
+		return err //return the info from the chain check error
+	}
+	//If we reach this, everything was checked without errors
+	return nil
+}
+
+func encodePrivateKeyCommon(privKeyBytes []byte) ([]byte, error) {
```
Comment (lohrke) [inline ubirch/test_benchmark_common_test.go:404]: Maybe rename this to something that makes it more clear that its a testing helper function and not a regular library function?

File: ubirch/test_benchmark_common_test.go [hunk 3]
```diff
@@ -244,3 +308,142 @@ func verifyUPPChain(t *testing.T, uppsArray [][]byte, startSignature []byte) err
...
+
+	x509Encoded, err := x509.MarshalECPrivateKey(privKey)
+	if err != nil {
+		return nil, err
+	}
+	pemEncoded := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: x509Encoded})
+	return pemEncoded, nil
+}
+
+func encodePublicKeyCommon(pubKeyBytes []byte) ([]byte, error) {
```
Comment (lohrke) [inline ubirch/test_benchmark_common_test.go:419]: Maybe rename this to something that makes it more clear that its a testing helper function and not a regular library function?

File: ubirch/test_benchmark_common_test.go [hunk 4]
```diff
@@ -244,3 +308,142 @@ func verifyUPPChain(t *testing.T, uppsArray [][]byte, startSignature []byte) err
...
+	x509EncodedPub, err := x509.MarshalPKIXPublicKey(pubKey)
+	if err != nil {
+		return nil, err
+	}
+	pemEncoded := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: x509EncodedPub})
+
+	return pemEncoded, nil
+}
+
+func decodePrivateKeyCommon(pemEncoded []byte) (*ecdsa.PrivateKey, error) {
```
Comment (lohrke) [inline ubirch/test_benchmark_common_test.go:436]: Maybe rename this to something that makes it more clear that its a testing helper function and not a regular library function?

File: ubirch/test_benchmark_common_test.go [hunk 5]
```diff
@@ -244,3 +308,142 @@ func verifyUPPChain(t *testing.T, uppsArray [][]byte, startSignature []byte) err
...
+	block, _ := pem.Decode(pemEncoded)
+	x509Encoded := block.Bytes
+	return x509.ParseECPrivateKey(x509Encoded)
+}
+
+func decodePublicKeyCommon(pemEncoded []byte) (*ecdsa.PublicKey, error) {
```
Comment (lohrke) [inline ubirch/test_benchmark_common_test.go:442]: Maybe rename this to something that makes it more clear that its a testing helper function and not a regular library function?

File: ubirch/ubirch_protocol.go [hunk 6]
```diff
@@ -19,21 +19,24 @@
...
 )
 
+// ProtocolType definition
 type ProtocolType uint8
 
 const (
-	Plain   ProtocolType = 0x00
-	Signed  ProtocolType = 0x22
-	Chained ProtocolType = 0x23
+	Plain   ProtocolType = 0x00 // Plain protocol, without hashing and signing
```
Comment (lohrke) [inline ubirch/ubirch_protocol.go:34]: This seems inconsistent across implementations, see e.g. here: https://github.com/ubirch/ubirch-protocol/blob/7a2b43be54745cba24e974a35be42a13b5f32b1a/ubirch/ubirch_protocol.h#L82 Plain type is 0x00 here and 0x01 in the c implementation. This needs to be 0x21.

File: ubirch/ubirch_protocol_test.go [hunk 7]
```diff
@@ -620,67 +1101,137 @@ func TestDecode(t *testing.T) {
...
-	log.Printf("pValue: %v", pValue)
+//TestECDSASignatureChanges tests if the signature of ECDSA changes for the same input using the top level protocol struct.
+//If it does not, there is likely a problem with the random number or nonce ('k') generation
+//which will allow attackers to calculate the private key from signatures.
+//Since only a 'small' number of signatures is checked this will most likely detect only 'total' failures
+//in k/nonce generation. See also  https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm#Security
+//For testing, 'signed' type UPPs are used. Both "new context for each UPP" and "one context across all UPPs" cases are tested
+func TestECDSASignatureChanges(t *testing.T) {
+	const nrOfSigsToCheck = 1000 //effective number of tests is two time this number as both "new context for each test" and "consistent context" cases are tested
+	userDataBytes := []byte("Hello World!")
```
Comment (lohrke) [inline ubirch/ubirch_protocol_test.go:1195]: Maybe better to use default input data global constant here for readability?

File: main/main.go [hunk 8]
```diff
@@ -34,7 +34,7 @@ import (
 
 func saveProtocolContext(p *ubirch.Protocol) error {
 	contextBytes, _ := json.Marshal(p)
-	err := ioutil.WriteFile("protocol.json", contextBytes, 444)
+	err := ioutil.WriteFile("../protocol.json", contextBytes, 444)
```
[the remark this answers is not in the document]
Comment (gruenwaldi) [reply main/main.go:37]: fixed

File: main/main.go [hunk 9]
```diff
@@ -45,7 +45,7 @@ func saveProtocolContext(p *ubirch.Protocol) error {
 }
 
 func loadProtocolContext(p *ubirch.Protocol) error {
-	contextBytes, err := ioutil.ReadFile("protocol.json")
+	contextBytes, err := ioutil.ReadFile("../protocol.json")
```
[the remark this answers is not in the document]
Comment (gruenwaldi) [reply main/main.go:48]: fixed

File: ubirch/crypto_test.go [hunk 10]
```diff
@@ -190,22 +192,33 @@ func TestCryptoContext_GenerateKey_NOTRDY(t *testing.T) {
...
 	asserter.Errorf(p.GenerateKey(defaultName, id), "Generating key without id")
 	pubKeyBytes, err = p.GetPublicKey(name)
-	asserter.Nilf(err, "Getting Public without name failed")
-	asserter.Nilf(pubKeyBytes, "Public Key for existing Key empty")
-
+	asserter.Errorf(err, "Getting Public without uuid")
+	asserter.Nilf(pubKeyBytes, "Public Key without uuid not empty")
+	privKeyBytes, err = getPrivateKey(context, name)
+	asserter.Errorf(err, "Getting Private Key without uuid")
+	asserter.Nilf(privKeyBytes, "Private Key without uuid not empty")
```
[the remark this answers is not in the document]
Comment (gruenwaldi) [reply ubirch/crypto_test.go:221]: fixed

File: ubirch/crypto_test.go [hunk 11]
```diff
@@ -233,29 +246,341 @@ func TestCryptoContext_GetPublicKey(t *testing.T) {
 	pubKeyBytesNew, err := p.GetPublicKey(defaultName)
 	asserter.NoError(err, "Getting Public key failed")
 	asserter.NotNilf(pubKeyBytesNew, "Public Key for existing Key empty")
+	asserter.Equal(lenPubkeyECDSA, len(pubKeyBytesNew), "len(public key) not correct a public key")
```
[the remark this answers is not in the document]
Comment (gruenwaldi) [reply ubirch/crypto_test.go:249]: fixed

File: ubirch/crypto_test.go [hunk 12]
```diff
@@ -233,29 +246,341 @@ func TestCryptoContext_GetPublicKey(t *testing.T) {
...
+	// asserter := assert.New(t)
+	// var context = &CryptoContext{
+	// 	Keystore: NewEncryptedKeystore([]byte(defaultSecret)),
+	// 	Names:    map[string]uuid.UUID{},
+	// }
+	// p := Protocol{Crypto: context, Signatures: map[uuid.UUID][]byte{}}
+	// certificate, err := p.GetCSR(defaultName)
+	// asserter.Nilf(err, "Getting CSR failed")
+	// asserter.NotNilf(certificate, "The Certificate is \"Nil\"")
+	// t.Errorf("not implemented")
```
[the remark this answers is not in the document]
Comment (gruenwaldi) [reply ubirch/crypto_test.go:308]: fixed

File: ubirch/crypto_test.go [hunk 13]
```diff
@@ -233,29 +246,341 @@ func TestCryptoContext_GetPublicKey(t *testing.T) {
...
 	asserter := assert.New(t)
+	requirer := require.New(t)
 	var context = &CryptoContext{
-		Keystore: NewEncryptedKeystore([]byte("2234567890123456")),
-		Names:    map[string]uuid.UUID{},
+		Keystore: &EncryptedKeystore{
+			Keystore: &keystore.Keystore{},
+			Secret:   []byte(defaultSecret),
+		},
+		Names: map[string]uuid.UUID{},
```
[the remark this answers is not in the document]
Comment (lohrke) [reply ubirch/crypto_test.go:572]: Should also be fixed for other occurrences.

[the remark this answers is not in the document]
Comment (gruenwaldi) [reply ubirch/crypto_test.go:572]: fixed for this and all other occurances

File: ubirch/keystore_test.go [hunk 14]
```diff
@@ -1,59 +1,252 @@
...
+	privKeyString := hex.EncodeToString(privKeyBytes)
+	asserter.Equalf(defaultPriv, privKeyString, "not equal")
+}
 
-	// can store key
-	if err := ks.SetKey(keyName, keyValue); err != nil {
-		t.Errorf("Error setting key %q: %s", keyName, err)
+// test, where the get method should fail
+func TestEncryptedKeystore_GetKeyFails_NOTRDY(t *testing.T) {
+
```
[the remark this answers is not in the document]
Comment (gruenwaldi) [reply ubirch/keystore_test.go:90]: fixed

File: ubirch/test_benchmark_common_test.go (hunk 2 above)
Comment (gruenwaldi) [reply ubirch/test_benchmark_common_test.go:404]: fixed

File: ubirch/test_benchmark_common_test.go (hunk 3 above)
Comment (gruenwaldi) [reply ubirch/test_benchmark_common_test.go:419]: fixed

File: ubirch/test_benchmark_common_test.go (hunk 4 above)
Comment (gruenwaldi) [reply ubirch/test_benchmark_common_test.go:436]: fixed

File: ubirch/test_benchmark_common_test.go (hunk 5 above)
Comment (gruenwaldi) [reply ubirch/test_benchmark_common_test.go:442]: fixed

File: ubirch/ubirch_protocol.go (hunk 6 above)
Comment (gruenwaldi) [reply ubirch/ubirch_protocol.go:34]: fixed

File: ubirch/ubirch_protocol_test.go (hunk 7 above)
Comment (gruenwaldi) [reply ubirch/ubirch_protocol_test.go:1195]: fixed

**Comment by lohrke** [comment]
@gruenwaldi @LaRox11 @ailox This is still open. Is there some reason for this because of other implementations which use this lib or did we just not to do the merge yet?

**Pull request merged**